Add resolver for custom snapshots paths

viddo opened the pull request:
## Summary

Provides a solution to #1650, by adding a configuration option to define a custom module that can resolve test<->snapshot paths. This enables people to decide where the snapshot files gets written on disk. 

The implementation was based on this suggestion: https://github.com/facebook/jest/issues/1650#issuecomment-247163522

I left the code to still assume `.snap` extension on the snapshot files however, since I wasn't convinced how useful it would be to override that too. Could be fairly easily solved by consolidating the `EXTENSION` usages.

## Motivation
My primary motivation for this PR stems from constraints by the [Bazel](https://bazel.build/) build system, as also explained on a [comment on _Add support for --preserve-symlinks_](https://github.com/facebook/jest/issues/5356#issuecomment-399246274):
> _"Bazel runs build tools and binaries inside sandboxes, temporary directories …"_

The sandbox is discarded after the test run, so any written files that are wanted afterwards must be written elsewhere (special directories also provided by Bazel, as it doesn't allow to write wherever you want outside the sandbox).

Judging by the reactions there are probably also other use-cases and rationales that motivates this feature.

## Test plan

Existing tests covers the default resolver behavior well. I've added some new unit and integration tests to cover the configurable part. 

I could add some more test cases of course, but wanted to get some initial feedback before spending more time on this.

SimenB commented:
CI is failing, mind looking into that test?

viddo commented:
Yes, will do. Haven't managed to reproduce with neither `yarn test`, `yarn run test-ci`, or `yarn run test-ci-partial` yet though, it passes locally (Mac OSX 11.12, Node v8.11), so I guess something be something else that differs. Will continue to investigate...

rickhanlonii commented on website/versioned_docs/version-22.4/Configuration.md:
```diff
@@ -725,6 +725,30 @@ monkey-patching the jasmine API. If you wanted to add even more jasmine plugins
 to the mix (or if you wanted some custom, project-wide matchers for example),
 you could do so in this module.
 
+### `snapshotResolver` [string]
```
This should not be added to the 22.4 docs as it will not land in 22.4

thymikee commented on packages/jest-snapshot/src/snapshot_resolver.js:
```diff
@@ -0,0 +1,67 @@
+      const snapshotPath = custom.resolveSnapshotPath(testPath, DOT_EXTENSION);
+      return snapshotPath.endsWith(DOT_EXTENSION)
+        ? snapshotPath
+        : snapshotPath + DOT_EXTENSION;
```
I think falling back to appending `DOT_EXTENSION` is not necessary. We shouldn't provide a layer of indirection in snapshot resolution logic but rely on user's implementation, just like in `resolveTestPath` below.

thymikee commented on packages/jest-snapshot/src/snapshot_resolver.js:
```diff
@@ -0,0 +1,67 @@
+
+  if (typeof custom.resolveSnapshotPath !== 'function') {
+    throw new TypeError(
+      'snapshotResolver does not have a resolveSnapshotPath function',
```
How about:
```
Custom snapshot resolver must implement a `resolveSnapshotPath` function.

Documentation: https://facebook.github.io/jest/docs/en/configuration.html#snapshotResolver
```

sidferreira commented:
@viddo What if, instead of picking a resolver to the snapshots, the resolver resolves the `Snapshot` module? By default returning `JestSnapshot` but allowing us to override all the `JestSnapshot` at our own risks?

viddo commented:
> @viddo What if, instead of picking a resolver to the snapshots, the resolver resolves the `Snapshot` module? By default returning `JestSnapshot` but allowing us to override all the `JestSnapshot` at our own risks?

@sidferreira, I'm not sure what you refer to by _`Snapshot` module_ and _`JestSnapshot`_. Are you referring to the whole jest-snapshot package export? The `SnapshotState` class? Something else? 

Either way, I'm reluctant to expose more internals than necessary, as it makes it harder to change those in the future as they practically then becomes a "public API" and affects the semver. 

Also, while I don't mind changing the implementation, the core team have the final say so I leave it to them to answer this.

SimenB commented on integration-tests/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,50 @@
+
+const snapshotDir = path.resolve(
+  __dirname,
+  path.join('..', 'snapshot-resolver', '__snapshots__'),
```
`path.resolve` normalizes, so just `path.resolve(__dirname, '../snapshot-resolver/__snapshots__');` should be enough

SimenB commented on integration-tests/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,50 @@
+
+const fileExists = filePath => {
+  try {
+    return fs.statSync(filePath).isFile();
```
`fs.existsSync` is not deprecated, please use it (async _is_ deprecated)

SimenB commented on integration-tests/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,50 @@
+
+describe('Custom snapshot resolver', () => {
+  const cleanup = () => {
+    [snapshotFile].forEach(file => {
```
should just drop the array here

SimenB commented on integration-tests/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,50 @@
+
+    // $FlowFixMe dynamic require
+    const content = require(snapshotFile);
+    expect(content['snapshots are written to custom location 1']).not.toBe(
```
`expect(content['snapshots are written to custom location 1']).toBeDefined()`

SimenB commented on packages/jest-snapshot/src/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,100 @@
+'use strict';
```
genereal comment, `path.resolve` normalizes, so you don't need to comma separate the varargs

viddo replied:
👍 better now?

viddo replied:
…realized I only fixed it on the other files, also updated this one now.

SimenB commented on packages/jest-snapshot/src/snapshot_resolver.js:
```diff
@@ -0,0 +1,66 @@
+
+function mustImplement(functionName: string) {
+  return (
+    `Custom snapshot resolver must implement a \`${functionName}\` function.\n\n` +
```
should the heading be bold?

SimenB commented:
Also, please rebase to ensure the snapshots are correct. I have a feeling some trailing newlines are gone, so it would fail when merged to master

viddo commented:
Will do, probably will have time to address the review comments tomorrow (Wednesday). Thanks for the additional review!

viddo commented:
> This needs an entry on the changelog

Good call ✅

> Also, please rebase to ensure the snapshots are correct. I have a feeling some trailing newlines are gone, so it would fail when merged to master

Gotcha. Rebased but didn't see any trouble, but probably better safe than sorry. :)

viddo commented:
Friendly bump; I believe I've addressed all the code review comments that's been brought up so far, what's expected to be changed/done for this PR to be accepted at this point?

re: the changelog, it keeps getting conflicts due to other updates, so I'll fix that as a last task.

gardner commented:
We are targeting web, ios, and android with a single code base. Having this feature will allow us to share more code between the different platforms.

crobinson42 commented:
Would love to see this moved forward 🌮

viddo commented:
Yay, thanks for approval!
Just came home from vacation and catching up on things, will try to get
this rebased and ready to be merged this week.

On Fri, Jul 27, 2018, 12:17 Cory Robinson <notifications@github.com> wrote:

> Would love to see this moved forward 🌮
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/jest/pull/6143#issuecomment-408514537>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA7uHbi5zM7zAYKQMkPfnF8-DvIrovaZks5uK2dLgaJpZM4T0JRZ>
> .
>

sppatel commented:
Any update on this? I am interested in this as well.

viddo commented:
Finally got around to this today; Just rebased and resolved conflicts.

From what I could see running the tests locally there are at least two tests that also fails on master:

```bash
 ● SearchSource › testPathsMatching › finds tests matching a pattern via testRegex

    expect(received).toEqual(expected)

    Expected value to equal:
      [".hiddenFolder/not-really-a-test.txt", "__testtests__/not-really-a-test.txt"]
    Received:
      ["__testtests__/not-really-a-test.txt"]

    Difference:

    - Expected
    + Received

      Array [
    -   ".hiddenFolder/not-really-a-test.txt",
        "__testtests__/not-really-a-test.txt",
      ]

      at findMatchingTests.then.data (packages/jest-cli/src/__tests__/SearchSource.test.js:123:26)

  ● SearchSource › testPathsMatching › finds tests matching a pattern via testMatch

    expect(received).toEqual(expected)

    Expected value to equal:
      [".hiddenFolder/not-really-a-test.txt", "__testtests__/not-really-a-test.txt"]
    Received:
      ["__testtests__/not-really-a-test.txt"]

    Difference:

    - Expected
    + Received

      Array [
    -   ".hiddenFolder/not-really-a-test.txt",
        "__testtests__/not-really-a-test.txt",
      ]

      at findMatchingTests.then.data (packages/jest-cli/src/__tests__/SearchSource.test.js:147:26)
```

viddo commented:
…appears the previously mentioned tests pass on CI, at least.

https://app.netlify.com/sites/jest-preview/deploys also appears to have been broken consistently as of late?

SimenB commented on e2e/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,39 @@
+
+    // $FlowFixMe dynamic require
+    const content = require(snapshotFile);
+    expect(content['snapshots are written to custom location 1']).toBeDefined();
```
`expect(content).toHaveProperty('snapshots are written to custom location 1');`

SimenB commented on e2e/__tests__/snapshot_resolver.test.js:
```diff
@@ -0,0 +1,39 @@
+  it('Resolves snapshot files using custom resolver', () => {
+    const result = runJest('snapshot-resolver', ['-w=1', '--ci=false']);
+
+    const info = result.stderr.toString();
```
you don't need the `toString` call

SimenB commented:
Mind fixing the lint error?

viddo commented:
Yeah. 
Note that the new test failure on [ci/circleci: test-node-8](https://circleci.com/gh/facebook/jest/30502?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) is also happening on other PRs so not unique to the changes here.

SimenB commented on CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 - `[pretty-format]` Support serializing `DocumentFragment` ([#6705](https://github.com/facebook/jest/pull/6705))
 - `[jest-validate]` Add `recursive` and `recursiveBlacklist` options for deep config checks ([#6802](https://github.com/facebook/jest/pull/6802))
 - `[jest-cli]` Check watch plugins for key conflicts ([#6697](https://github.com/facebook/jest/pull/6697))
+- `[jest-snapshot]` Enable configurable snapshot paths ([#6143](https://github.com/facebook/jest/pull/6143))
```
should be under "master"

wallynm commented:
Waiting so much for this update! Good job guys! <3

MQuy commented:
Any updates on this pr 🤔

viddo commented:
@MQuy From my end I'm rebasing the PR every now and then to help it stay mergeable. 

For now I have worked around this problem by some external scripts that move/rename the snapshots files after the test run (due to the problems of https://github.com/facebook/jest/issues/5356#issuecomment-399246274 for the curious ones; I'll update the PR desc since I forgot to do that earlier I realize now).

As Jest is a very popular testing tool and there are a lot of other open issues and PRs I suspect the maintainers might be a bit overwhelmed with more important stuff right now.

I'd guess that anyone that also wants this feature can help "boost it" by adding a 👍 / ❤️ / 🎉  reaction to the original post at the top, as that tend to give some indication of popularity and might help bump it up the sort list of [open PRs](https://github.com/facebook/jest/pulls?q=is%3Apr+is%3Aopen+sort%3Areactions-%2B1-desc).

SimenB commented:
@viddo needs a quick rebase/merge for the changelog entry to be in the correct place.

---

@cpojer @aaronabramov @mjesun can one of you make a decision on this? I'm 🤷‍♂️on it personally, FWIW. While I'm not a fan of more configuration options (I refer to https://fishshell.com/docs/current/design.html often, thanks for that link!) there seems to be a lot of people wanting this feature and it does solve real use cases that I don't think makes sense to solve generally within Jest core.

I like [Ava's approach](https://github.com/avajs/ava/blob/bfe16300e07c12a1bad2e4f02898b1e7486daf70/readme.md#snapshot-testing) of getting a name for the snapshot directory based on the name/location of the test file, but if we were to copy that approach we still wouldn't solve the use case posed in the OP. Ava does support a `snapshotDir` option as well, though.

cpojer commented:
Yeah, I'm gonna fight against this any longer. I would like Jest's config normalization to call both of the functions to ensure that they resolve to each other, like `resolveTestPath(resolveSnapshotPath(theoreticalTestPath))` or something like it, and throw if they are not correct, to make sure that people use the feature correctly and we don't create weird inconsistencies.

ljharb commented:
(the implication of message is that you’re *not* going to fight any longer? I’ll delete this message after a clarification)

SimenB commented on packages/jest-editor-support/src/Snapshot.js:
```diff
@@ -12,7 +12,8 @@
 import {getASTfor} from './parsers/babylon_parser';
-import {utils} from 'jest-snapshot';
+import {buildSnapshotResolver, utils} from 'jest-snapshot';
+import type {ProjectConfig} from 'types/Config';
```
We generally have type import at the top on their own

SimenB commented on packages/jest-snapshot/src/snapshot_resolver.js:
```diff
@@ -0,0 +1,69 @@
+    throw new TypeError(mustImplement('resolveTestPath'));
+  }
+
+  return {
```
above here is probably a good place for the `fakeTestpath === custom.resolveTestPath(custom.resolveSnapshotPath(fakeTestpath))` check

SimenB commented:
@viddo mind fixing the comment? Then we can merge 🙂

Joonpark13 commented:
Just adding my +1: Would love to see this get merged!

SimenB commented on packages/jest-snapshot/src/snapshot_resolver.js:
```diff
@@ -67,3 +71,21 @@ function mustImplement(functionName: string) {
+  const transformedPath = custom.resolveTestPath(
+    custom.resolveSnapshotPath(fakeTestPath),
+  );
+  if (transformedPath !== fakeTestPath) {
```
You can pass these 2 through `slash` to normalize the slashes

SimenB replied:
I just did the thing I suggested in the edit, seems straight forward enough.

🤞 for CI

SimenB commented:
Nice job, thanks for sticking with us!

nicoabie commented:
@viddo thanks for the great work, it is very useful indeed!!! and @SimenB for the reviews an merging.

I'm using it and works very well except for one little detail. Both testPath and snapshotFilePath args are absolute paths. I think it would be desirable they be relative to the CWD.

In the meantime, this is a hack what works around the issue.

```javascript
module.exports = {
  // resolves from test to snapshot path
  resolveSnapshotPath: (testPath, snapshotExtension) =>
    '__snapshots__/' + /src\/(.*)/.exec(testPath)[1] + snapshotExtension,

  // resolves from snapshot to test path
  resolveTestPath: (snapshotFilePath, snapshotExtension) =>
    ('src/' + /__snapshots__\/(.*)/.exec(snapshotFilePath)[1]).slice(0, -snapshotExtension.length),

  // Example test path, used for preflight concistency check of the implementation above
  testPathForConsistencyCheck: 'src/components/example.test.js',
};
```

SimenB commented:
It might make sense to pass in `config`, then you can do `path.relative(config.rootDir, testPath)` if you want?

nicoabie commented:
I think there isn't any use case where you will need absolute paths so users will always end up transforming to relative.

The absolute path of any file will always differ between different computers, so for the same test file you will end up having a different snapshot per user.

SimenB commented:
>I think there isn't any use case where you will need absolute paths so users will always end up transforming to relative.

If you need to read the file or something. IDK, but it's easier to say "here is the file" and let people do whatever they want with the path rather than making it relative.

If we make it relative, what should we make it relative to? `cwd`? `rootDir`? Your config file?

nicoabie commented:
I may be wrong but I can only imagine "src" folder. 
And I know it is a convention to call it that way.

But I believe that is the more common use case and should be the default. Your idea about passing in a config is great but should be for the other cases.

Does that makes sense?

SimenB commented:
Jest for instance does not use `src/`. Anyways, please open up a new issue for discussion here, no reason to have it in the PR and ping all who have contributed here 🙂

G-Rath commented:
@SimenB if you have a quick minute, could you confirm for me the version this was released in?

I can't find any mention of `snapshotResolver` in the master changelog.

I'm trying to get as much info as I can for the [typings](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34715) :)

SimenB commented:
24.0.0

>I can't find any mention of snapshotResolver in the master changelog, nor any blog posts.

Easiest is to search for the pr number in the changelog.

It's also mentioned in the release blog post: https://jestjs.io/blog/2019/01/25/jest-24-refreshing-polished-typescript-friendly#other-highlights

G-Rath commented:
> I don't see why it shouldn't work though

Thats what I thought, but Jest complains:

```

  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    C:\Users\G-Rath\workspace\projects-personal\strongly-typed-event-emitter\test\snapshotResolver.ts:3
      resolveSnapshotPath: (testPath: string, snapshotExtension: string) => {
                                    ^

    SyntaxError: Unexpected token :

      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:471:17)
      at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:513:25)

```

It works fine if I convert it to `.js`. This is my transformers:

```
    "transform": {
      ".(ts|tsx)": "ts-jest"
    }
```

I also tried putting in `ts-node`, but it changed nothing. From my very limited understanding of Jest, it looks like maybe it's doing `snapshotResolver` before it loads/handles transformers?

SimenB commented:
Oh sorry, I misread the question as resolving _to_ ts files, not using ts in a resolver.

That probably doesn't work, but we can (probably) make it work. Mind opening up a feature request?

G-Rath commented:
> Mind opening up a feature request?

Just what I wanted to hear ;)

I've made #8330 for this.

CaerusKaru commented:
@viddo I know it's been a while, but is there any chance you could provide the setup that you use to get this to work with Bazel? Might be helpful to document this in the [rules_nodejs repository example for jest integration](https://github.com/bazelbuild/rules_nodejs/tree/master/examples/jest).

longlho commented:
@CaerusKaru you're probably looking for https://github.com/formatjs/formatjs/blob/main/tools/jest.bzl#L46-L66 most things work out of the box including snapshot update
